### PR TITLE
MySQL VM new relations implementation

### PR DESCRIPTION
**Pull request by paulomach**
# Issue

MySQL charm must implement new relations

# Solution

Using data-platform-libs, implemented requires side of new relations.
Missing population of TLS related fields.

# Testing

Testing was implemented in unit tests, and manual validation was done with test charm.
Unit tests to follow in another PR.

# Release Notes

- `lib/charms/data_platform_libs/v0/database_provides.py` - updated from upstream
- `lib/charms/mysql/v0/mysql.py` - implemented new methods to support relation requirements
- `tests/unit/test_mysql.py` - utest for new methods
- `metadata.yaml` - add relation
- `src/charm.py` - instantiate relation, set workload version
- `tests/unit/test_charm.py` - fix for set workload version
- `src/constants.py` - trivial
- `src/relations/database.py` - implements new relation
- `tests/unit/test_database.py` - utest new relation

### CODE REVIEW

File: metadata.yaml [hunk 1]
```diff
@@ -24,6 +24,8 @@ provides:
     interface: mysql-router
   shared-db:
     interface: mysql-shared
+  database:
+    interface: database-client
```
Comment (marceloneppel) [inline metadata.yaml:28]:
```suggestion
    interface: mysql-client
```

File: tests/unit/test_database.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        self.harness.update_relation_data(
+            self.database_relation_id, "app", {"database": "test_db"}
+        )
+
+        # confirm that the relation databag is updated
+        database_relation_databag = self.harness.get_relation_data(
+            self.database_relation_id, self.harness.charm.app
+        )
+        database_relation = self.charm.model.get_relation(DB_RELATION_NAME)
+        app_unit = list(database_relation.units)[0]
```
Comment (marceloneppel) [inline tests/unit/test_database.py:73]: Is this code block duplicated? (L51-L52)

File: src/relations/database.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    def _get_or_set_password(self, relation) -> str:
+        """Retrieve password from cache or generate a new one.
+
+        Args:
+            relation (str): The relation for each the password is cached.
+
+        Returns:
+            str: The password.
+        """
+        if password := relation.data[self.charm.app].get("password"):
```
Comment (marceloneppel) [inline src/relations/database.py:52]: This syntax is interesting 😀

File: src/relations/database.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        db_pass = self._get_or_set_password(event.relation)
+        db_version = self.charm._mysql.get_mysql_version()
+
+        remote_app = event.app.name
+
+        try:
+            primary_endpoint = self.charm._mysql.get_cluster_primary_address()
+            self.database.set_credentials(relation_id, db_user, db_pass)
+            self.database.set_endpoints(relation_id, primary_endpoint)
+            self.database.set_version(relation_id, db_version)
```
Comment (marceloneppel) [inline src/relations/database.py:82]: Can this data updates cause issues if the later user creation fails? Maybe the application can try to create tables using the non created user. I don't remember exactly in which situations the relation data is committed (I guess it's committed if the hook exits without an error/exception).

**Review by marceloneppel — COMMENTED** [review]
Great work, Paulo! I left some comments about the implementation.

File: src/relations/database.py (hunk 4 above)
Comment (paulomach) [reply src/relations/database.py:82]: That's safe to do. It's only committed after successful hook execution.

File: tests/unit/test_database.py (hunk 2 above)
Comment (paulomach) [reply tests/unit/test_database.py:73]: I was checking if people were paying attention on reviews ;)

File: src/charm.py [hunk 5]
```diff
@@ -122,12 +125,16 @@ def _on_start(self, event: StartEvent) -> None:
...
             self._mysql.wait_until_mysql_connection()
+            workload_version = self._mysql.get_mysql_version()
+            self.unit.set_workload_version(workload_version)
         except MySQLConfigureMySQLUsersError:
             self.unit.status = BlockedStatus("Failed to initialize MySQL users")
             return
         except MySQLConfigureInstanceError:
             self.unit.status = BlockedStatus("Failed to configure instance for InnoDB")
             return
+        except MySQLClientError:
```
Comment (shayancanonical) [inline src/charm.py:136]: 1. our pattern has been to not catch `MySQLClientError` in the charm code
2. shouldnt the charm go into blockedstatus?

File: src/relations/database.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+            logger.info(f"Created user for app {remote_app}")
+        except MySQLCreateApplicationDatabaseAndScopedUserError:
+            logger.error(f"Failed to create scoped user for app {remote_app}")
+            self.charm.unit.status = BlockedStatus("Failed to create scoped user")
+            return
+        except MySQLClientError as e:
+            logger.exception("Failed to retrieve cluster info", exc_info=e)
+            self.charm.unit.status = BlockedStatus("Failed to retrieve cluster info")
+            return
```
Comment (shayancanonical) [inline src/relations/database.py:107]: do you need these `return`s? they're at the end of the branch for the method. same with line 124

Comment (paulomach) [reply src/relations/database.py:107]: not really

File: src/charm.py (hunk 5 above)
Comment (paulomach) [reply src/charm.py:136]: 1. Ok
2. not really. this is an optional field

**Review by shayancanonical — APPROVED** [review]

**Pull request merged**